File: prefig/core/vector.py
import lxml.etree as ET
import logging
import math
import numpy as np
import copy
from . import user_namespace as un
from . import utilities as util
from . import math_utilities as math_util
from . import arrow
from . import label

log = logging.getLogger('prefigure')

# Add a graphical element describing a vector
def vector(element, diagram, parent, outline_status):
    if outline_status == 'finish_outline':
        finish_outline(element, diagram, parent)
        return

    # v describes the mathematical vector (displacement),
    # which is scaled by scale, tail is the location of 
    # the tail
    try:
        v = un.valid_eval(element.get('v'))
    except:
        log.error(f"Error parsing vector attribute @v={element.get('v')}")
        return

    tail = un.valid_eval(element.get('tail', '[0,0]'))
    scale = un.valid_eval(element.get('scale', '1'))
    v = scale * v
    w = v + tail

    diagram.register_source_data(element, 'v', v)
    diagram.register_source_data(element, 'head', w)

    # Specify where we want the head to appear.  By default,
    # it's at the tip of the vector, but it can be anywhere
    # along the shaft at a location specified by 0 <= t <= 1
    t = element.get('head-location', None)
    if t is not None:
        t = float(t)
        head_loc = (1-t)*tail + t * w

    if diagram.output_format() == 'tactile':
        element.set('fill', 'black')
        element.set('stroke', 'black')
    else:
        util.set_attr(element, 'stroke', 'black')
        util.set_attr(element, 'fill', 'none')
    util.set_attr(element, 'thickness', '3')

    vector = ET.Element('path')
    diagram.add_id(vector, element.get('id'))
    diagram.register_svg_element(element, vector)
    util.add_attr(vector, util.get_2d_attr(element))

    # Now add the head using an SVG marker
    if t is not None:
        location = 'marker-mid'
    else:
        location = 'marker-end'
    arrow_id = arrow.add_arrowhead_to_path(
        diagram,
        location,
        vector,
        arrow_width=element.get('arrow-width', None),
        arrow_angles=element.get('arrow-angles', None)
    )

    # we need to pull the tip of the vector in a bit to accommodate
    # the arrowhead
    p0 = diagram.transform(tail)
    p1 = diagram.transform(w)
    diff = p1 - p0
    length = math_util.length(diff)
    angle = math.atan2(diff[1], diff[0])
    diagram.register_source_data(element, 'angle', angle)

    arrow_head_length = arrow.get_arrow_length(arrow_id)
    thickness = un.valid_eval(element.get('thickness'))
    if location == 'marker-end':
        length -= thickness * arrow_head_length
        p1 = length * np.array([math.cos(angle),math.sin(angle)]) + p0

    # Here is the shaft of the vector.  If the head is not at
    # the tip, we add a waypoint along the line where the head
    # will appear
    cmds = []

    cmds.append('M ' + util.pt2str(p0))
    if t is not None:
        cmds.append('L ' + util.pt2str(diagram.transform(head_loc)))
    cmds.append('L ' + util.pt2str(p1))
    d = ' '.join(cmds)
    vector.set('d', d)

    if outline_status == 'add_outline':
        diagram.add_outline(element, vector, parent)
        return

    if element.get('outline', 'no') == 'yes' or diagram.output_format() == 'tactile':
        diagram.add_outline(element, vector, parent)
        finish_outline(element, diagram, parent)
    else:
        original_parent = parent
        parent = add_label(element, diagram, parent)
        parent.append(vector)

        # no label has been added if the parent hasn't changed
        if original_parent == parent:
            diagram.register_svg_element(element, vector)
            return

        diagram.register_svg_element(element, parent)
        # if there is a label, then the id is on the outer <g> element
        # so we need to remove it from the children
        if element.get('id', 'none') == parent.get('id'):
            element.attrib.pop('id')
        for child in parent:
            if child.get('id', None) is not None:
                child.attrib.pop('id')

def finish_outline(element, diagram, parent):
    original_parent = parent
    parent = add_label(element, diagram, parent)

    # if we've added a label, remove the id's from element under the parent <g>
    if original_parent != parent:
        for child in parent:
            if child.get('id', None) is not None:
                child.attrib.pop('id')

    diagram.finish_outline(element,
                           element.get('stroke'),
                           element.get('thickness'),
                           element.get('fill', 'none'),
                           parent)

# dictionary for alignments and offsets based on half-quadrant
alignment_dict = {0:('se',-1),
                  1:('nw',1),
                  2:('ne',-1),
                  3:('sw',1),
                  4:('nw',-1),
                  5:('se',1),
                  6:('sw',-1),
                  7:('ne',1)}

def add_label(element, diagram, parent):
    # Is there a label associated with point?
    text = element.text

    # is there a label here?
    has_text = text is not None and len(text.strip()) > 0
    all_comments = all([subel.tag is ET.Comment for subel in element])
    if has_text or not all_comments:
        # If there's a label, we'll bundle the label and vector in a group
        group = ET.SubElement(parent, 'g')
        diagram.add_id(group, element.get('id'))

        # Now we'll create a new XML element describing the label
        el = copy.deepcopy(element)
        el.tag = 'label'

        # determine alignment of the label
        alignment = element.get('alignment', None)
        user_offset = element.get('offset', None)
        angle = diagram.get_source_data(element, 'angle')
        if alignment is None:
            angle_degrees = math.degrees(-angle)
            while angle_degrees < 0:
                angle_degrees += 360
            half_quadrant = math.floor(angle_degrees / 45)
            alignment, offset_dir = alignment_dict[half_quadrant]
            el.set('alignment', alignment)

            normal = (math.cos(-angle), math.sin(-angle))
            direction = offset_dir * np.array((-normal[1], normal[0]))
            offset = 4 * direction
            if user_offset is not None:
                offset += un.valid_eval(user_offset)
            el.set('abs-offset', util.np2str(offset))
        else:
            displacement = label.alignment_displacement[alignment]
            def_offset = np.array([4*(displacement[0]+0.5),
                                   4*(displacement[1]-0.5)])
            if user_offset is not None:
                def_offset += un.valid_eval(user_offset)
            el.set('offset', util.np2str(def_offset))

        head = diagram.get_source_data(element, 'head')
        el.set('anchor', util.pt2long_str(head, spacer=','))

        # add the label graphical element to the group
        label.label(el, diagram, group)
        return group
    else:
        return parent

File: packages/prefig-rust/prefig-core/src/core/vector.rs
//! Port of prefig/core/vector.py.

use crate::core::diagram::Diagram;
use crate::core::utilities::{self as util, np2str, pt2long_str, pt2str};
use crate::core::{arrow, label};
use crate::value::Value;
use crate::xml::{self, El};

pub fn vector(element: &El, diagram: &mut Diagram, parent: &El, outline_group: Option<&El>) {
    let v_attr = element.borrow().get("v");
    let Some(v) = v_attr
        .as_deref()
        .and_then(|a| diagram.ctx.valid_eval(a).ok())
        .and_then(|v| v.as_vec_f64().ok())
    else {
        log::error!("Error parsing vector attribute @v={v_attr:?}");
        return;
    };

    let tail_attr = element.borrow().get_or("tail", "[0,0]");
    let tail = diagram
        .ctx
        .valid_eval(&tail_attr)
        .ok()
        .and_then(|v| v.as_vec_f64().ok())
        .unwrap_or(vec![0.0, 0.0]);
    let scale_attr = element.borrow().get_or("scale-length", "1");
    let scale = diagram
        .ctx
        .valid_eval(&scale_attr)
        .ok()
        .and_then(|v| v.as_num().ok())
        .unwrap_or(1.0);
    let v = [scale * v[0], scale * v[1]];
    let tail = [tail[0], tail[1]];
    let w = [v[0] + tail[0], v[1] + tail[1]];

    let vec_value = |p: [f64; 2]| Value::Array(vec![Value::Num(p[0]), Value::Num(p[1])]);
    diagram.register_source_data(element, "v", vec_value(v));
    diagram.register_source_data(element, "head", vec_value(w));
    diagram.register_source_data(element, "tail", vec_value(tail));

    // where along the shaft the head appears (default: the tip)
    let t: Option<f64> = element
        .borrow()
        .get("head-location")
        .and_then(|s| s.parse().ok());
    let head_loc = t.map(|t| {
        [
            (1.0 - t) * tail[0] + t * w[0],
            (1.0 - t) * tail[1] + t * w[1],
        ]
    });

    if diagram.output_format() == "tactile" {
        element.borrow_mut().set("fill", "black");
        element.borrow_mut().set("stroke", "black");
    } else {
        util::set_attr(element, "stroke", "black", &mut diagram.ctx);
        util::set_attr(element, "fill", "none", &mut diagram.ctx);
    }
    util::set_attr(element, "thickness", "3", &mut diagram.ctx);

    let vector = xml::new_element("path");
    let attrs = util::get_2d_attr(element, diagram);
    util::add_attr(&vector, attrs);

    let location = if t.is_some() {
        "marker-mid"
    } else {
        "marker-end"
    };
    let arrow_width = element.borrow().get("arrow-width");
    let arrow_angles = element.borrow().get("arrow-angles");
    let arrow_id = arrow::add_arrowhead_to_path(
        diagram,
        location,
        &vector,
        arrow_width.as_deref(),
        arrow_angles.as_deref(),
    );

    // pull the tip in a bit to accommodate the arrowhead
    let p0 = diagram.transform(tail);
    let mut p1 = diagram.transform(w);
    let diff = [p1[0] - p0[0], p1[1] - p0[1]];
    let mut length = (diff[0] * diff[0] + diff[1] * diff[1]).sqrt();
    let angle = diff[1].atan2(diff[0]);
    diagram.register_source_data(element, "angle", Value::Num(angle));

    let arrow_head_length = arrow_id
        .and_then(|id| diagram.arrow_lengths.get(&id).copied())
        .unwrap_or(0.0);
    let thickness_attr = element.borrow().get_or("thickness", "3");
    let thickness = diagram
        .ctx
        .valid_eval(&thickness_attr)
        .ok()
        .and_then(|v| v.as_num().ok())
        .unwrap_or(3.0);
    if location == "marker-end" {
        length -= thickness * arrow_head_length;
        p1 = [
            length * angle.cos() + p0[0],
            length * angle.sin() + p0[1],
        ];
    }

    let mut cmds = vec![format!("M {}", pt2str(p0, " "))];
    if let Some(head_loc) = head_loc {
        cmds.push(format!("L {}", pt2str(diagram.transform(head_loc), " ")));
    }
    cmds.push(format!("L {}", pt2str(p1, " ")));
    vector.borrow_mut().set("d", &cmds.join(" "));

    let mut parent = parent.clone();
    if label::has_label(element) {
        let group = xml::sub_element(&parent, "g");
        let id = element.borrow().get("id");
        diagram.add_id(&group, id.as_deref());
        diagram.register_svg_element(element, &group);
        parent = group;
        add_label(element, diagram, &parent);
    } else {
        let id = element.borrow().get("id");
        diagram.add_id(&vector, id.as_deref());
        diagram.register_svg_element(element, &vector);
    }

    if let Some(outline_group) = outline_group {
        diagram.add_outline(element, &vector, outline_group, None);
        finish_outline(element, diagram, &parent);
    } else if element.borrow().get_or("outline", "no") == "yes"
        || diagram.output_format() == "tactile"
    {
        diagram.add_outline(element, &vector, &parent, None);
        finish_outline(element, diagram, &parent);
    } else {
        xml::append(&parent, &vector);
    }
}

fn finish_outline(element: &El, diagram: &mut Diagram, parent: &El) {
    let stroke = element.borrow().get("stroke");
    let thickness = element.borrow().get("thickness");
    let fill = element.borrow().get_or("fill", "none");
    diagram.finish_outline(element, stroke, thickness, &fill, parent);
}

/// alignment and offset direction by half-quadrant of the vector's angle
fn alignment_for_half_quadrant(hq: i64) -> (&'static str, f64) {
    match hq {
        0 => ("se", -1.0),
        1 => ("nw", 1.0),
        2 => ("ne", -1.0),
        3 => ("sw", 1.0),
        4 => ("nw", -1.0),
        5 => ("se", 1.0),
        6 => ("sw", -1.0),
        _ => ("ne", 1.0),
    }
}

fn add_label(element: &El, diagram: &mut Diagram, parent: &El) {
    let el = xml::deep_copy(element);
    el.borrow_mut().tag = "label".to_string();

    let alignment = element.borrow().get("alignment");
    let user_offset = element.borrow().get("offset");
    let angle = diagram
        .get_source_data(element, "angle")
        .and_then(|v| v.as_num().ok())
        .unwrap_or(0.0);

    match alignment {
        None => {
            let mut angle_degrees = (-angle).to_degrees();
            while angle_degrees < 0.0 {
                angle_degrees += 360.0;
            }
            let half_quadrant = (angle_degrees / 45.0).floor() as i64;
            let (alignment, offset_dir) = alignment_for_half_quadrant(half_quadrant);
            el.borrow_mut().set("alignment", alignment);

            let normal = [(-angle).cos(), (-angle).sin()];
            let direction = [offset_dir * -normal[1], offset_dir * normal[0]];
            let mut offset = [4.0 * direction[0], 4.0 * direction[1]];
            if let Some(user_offset) = &user_offset {
                if let Some(v) = diagram
                    .ctx
                    .valid_eval(user_offset)
                    .ok()
                    .and_then(|v| v.as_vec_f64().ok())
                {
                    offset = [offset[0] + v[0], offset[1] + v[1]];
                }
            }
            el.borrow_mut().set("abs-offset", &np2str(offset));
        }
        Some(alignment) => {
            let displacement =
                label::alignment_displacement(&alignment).unwrap_or([-0.5, 0.5]);
            let mut def_offset = [
                4.0 * (displacement[0] + 0.5),
                4.0 * (displacement[1] - 0.5),
            ];
            if let Some(user_offset) = &user_offset {
                if let Some(v) = diagram
                    .ctx
                    .valid_eval(user_offset)
                    .ok()
                    .and_then(|v| v.as_vec_f64().ok())
                {
                    def_offset = [def_offset[0] + v[0], def_offset[1] + v[1]];
                }
            }
            el.borrow_mut().set("offset", &np2str(def_offset));
        }
    }

    let label_location = element
        .borrow()
        .get("label-location")
        .and_then(|s| diagram_eval_num(diagram, &s))
        .map(|l| l.clamp(0.0, 1.0))
        .unwrap_or(1.0);
    let head = diagram
        .get_source_data(element, "head")
        .and_then(|v| v.as_vec_f64().ok())
        .unwrap_or(vec![0.0, 0.0]);
    let tail = diagram
        .get_source_data(element, "tail")
        .and_then(|v| v.as_vec_f64().ok())
        .unwrap_or(vec![0.0, 0.0]);
    let anchor = [
        label_location * head[0] + (1.0 - label_location) * tail[0],
        label_location * head[1] + (1.0 - label_location) * tail[1],
    ];
    el.borrow_mut().set("anchor", &pt2long_str(anchor, ","));

    label::label(&el, diagram, parent, None);
}

fn diagram_eval_num(diagram: &mut Diagram, s: &str) -> Option<f64> {
    diagram.ctx.valid_eval(s).ok()?.as_num().ok()
}
